Modules/selectmodule.c

```diff
@@ -1294,15 +1284,23 @@ newPyEpoll_Object(PyTypeObject *type, int sizehint, SOCKET fd)
 }
 
 
+/*[clinic input]
+@classmethod
+select.epoll.__new__
+
+    sizehint: int(c_default="FD_SETSIZE - 1") = -1
```

Comment:
Negative sizehint is an error.

Reply:
Should it be an error? The tests only check that `-2` should raise ValueError, not `-1`.  The existing doc-string mentions that `-1` should mean the default size, though it currently doesn't.

Perhaps we should fix the code to special-case `-1`? Currently the only way to get the default is to not specify `sizehint`, but that's one type of inconsistency AC aims to remove.

Reply:
See the check at line 1304. The current code treats -1 as invalid value. Argument Clinic conversion shouldn't change semantic (too much). If there is a bug in the current code, fix it in a separate issue, so that the fix can be backported.

Reply:
This may not actually matter much. From the docs: "*sizehint* and *flags* are deprecated and completely ignored." Looking at the code, this is true for `flags`, but `sizehint` is actually passed to `epoll_create` if `HAVE_EPOLL_CREATE1` isn't defined.

So there are two problems right now:
1. The current doc-string is simply wrong: "sizehint must be a positive integer or -1 for the default size"
2. Since `-1` is actually invalid, there's currently no value that can be used in the doc-string to describe the default for `sizehint`, which is slightly bad for AC conversion.

Fixing the doc-string to describe the actual behavior seems like the way to go. I'm just not sure what that is regarding `sizehint`: is it actually ignored? I.e. are the docs wrong on this?

I'm really unsure what to do regarding the default value for `sizehint` shown in the doc-string.

Reply:
@serhiy-storchaka Could you help decide how to resolve this point?

Reply:
I would open a separate issue for restoring the *sizehint* behavior before 2fb9ae9dfc5a6830d902332ba93b13387e292ddb.

Reply:
I'm summarizing in hope to move forward with this.

The default for `sizehint` should ideally be `FD_SETSIZE - 1`, but `FD_SETSIZE` isn't exposed at the Python level. The current behavior sets `sizehint` to `FD_SETSIZE - 1` if not supplied, but that's a type of inconsistency that AC aims to remove.

`-1` is an invalid value, therefore it should probably not be used as a default value since that would change the existing semantics too much.

The docs say that `sizehint` is deprecated and ignored, in which case this discussion would be moot. However in the code `sizehint` is still passed to `epoll_create()` if `epoll_create1()` is not available:

```c
#ifdef HAVE_EPOLL_CREATE1
        self->epfd = epoll_create1(EPOLL_CLOEXEC);
#else
        self->epfd = epoll_create(sizehint);
#endif
```

Reply:
-1 was valid before 2fb9ae9dfc5a6830d902332ba93b13387e292ddb (but 0 was not valid). I'm not sure this change is intentional. It is better to open a separate issue for this.

Reply:
@serhiy-storchaka, I'll open a new issue. In the meantime, what do we do with the default value of `sizehint` in this branch?

Reply:
Depending on the results of that issue. Likely -1 will be the default value.

Reply:
Created bpo-32568.